### PR TITLE
Direction: Rename `from_normalized` to `new_unchecked`

### PR DESCRIPTION
**Pull request by doonv**
# Objective

`Direction2d::from_normalized` & `Direction3d::from_normalized` don't emphasize that importance of the vector being normalized enough.

## Solution

Rename `from_normalized` to `new_unchecked` and add more documentation.

---

`Direction2d` and `Direction3d` were added somewhat recently in https://github.com/bevyengine/bevy/pull/10466 (after 0.12), so I don't think documenting the changelog and migration guide is necessary (Since there is no major previous version to migrate from).

But here it is anyway in case it's needed:

## Changelog

- Renamed `Direction2d::from_normalized` and `Direction3d::from_normalized` to `new_unchecked`.

## Migration Guide

- Renamed `Direction2d::from_normalized` and `Direction3d::from_normalized` to `new_unchecked`.

### CODE REVIEW

File: crates/bevy_math/src/primitives/dim2.rs [hunk 1]
```diff
@@ -24,6 +24,18 @@ impl Direction2d {
         Self::new_and_length(value).map(|(dir, _)| dir)
     }
 
+    /// Create a [`Direction2d`] from a [`Vec2`] that is already normalized.
+    ///
+    /// ## Safety
```
Comment (tguichaoua) [inline crates/bevy_math/src/primitives/dim2.rs:29]:
```suggestion
    /// # Safety
```

File: crates/bevy_math/src/primitives/dim3.rs [hunk 2]
```diff
@@ -28,6 +28,18 @@ impl Direction3d {
         Self::new_and_length(value).map(|(dir, _)| dir)
     }
 
+    /// Create a [`Direction3d`] from a [`Vec3`] that is already normalized.
+    ///
+    /// ## Safety
```
Comment (tguichaoua) [inline crates/bevy_math/src/primitives/dim3.rs:33]:
```suggestion
    /// # Safety
```

File: crates/bevy_math/src/primitives/dim2.rs [hunk 3]
```diff
@@ -24,6 +24,18 @@ impl Direction2d {
...
+    ///
+    /// This function only checks if `value` is normalized on debug builds,
+    /// release builds can allow the creation of a [`Direction2d`] that is not normalized.
```
Comment (tguichaoua) [inline crates/bevy_math/src/primitives/dim2.rs:32]:
```suggestion
    /// `value` must be normalized, i.e it's length must be `1.0`.
```
I think it's better to have something more straightforward for the safety section.

File: crates/bevy_math/src/primitives/dim3.rs [hunk 4]
```diff
@@ -28,6 +28,18 @@ impl Direction3d {
...
+    ///
+    /// This function only checks if `value` is normalized on debug builds,
+    /// release builds can allow the creation of a [`Direction3d`] that is not normalized.
```
Comment (tguichaoua) [inline crates/bevy_math/src/primitives/dim3.rs:36]:
```suggestion
    /// `value` must be normalized, i.e it's length must be `1.0`.
```
Same here

File: crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs [hunk 5]
```diff
@@ -166,7 +166,8 @@ impl Bounded3d for Capsule {
     fn aabb_3d(&self, translation: Vec3, rotation: Quat) -> Aabb3d {
         // Get the line segment between the hemispheres of the rotated capsule
         let segment = Segment3d {
-            direction: Direction3d::from_normalized(rotation * Vec3::Y),
+            // SAFETY: Multipling a normalized vector (Vec3::Y) with a rotation returns a normalized vector.
+            direction: unsafe { Direction3d::new_unchecked(rotation * Vec3::Y) },
```
Comment (tguichaoua) [inline crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs:170]: Can `Direction(3d,2d)` implements rotation ?  (probably in another PR)
to directly use `rotation * Direction3d::Y` without unsafe.

**Review by tguichaoua — CHANGES_REQUESTED** [review]

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Conventionally, `unsafe` in Rust is used *strictly* for memory safety issues. While I strongly think that there should be alternatives that work like `unsafe`, but for other forms of danger, we should stick to the conventions of the ecosystem.

That said, I love the rest of these changes and think we need to warn more aggressively here.

File: crates/bevy_math/src/primitives/dim2.rs [hunk 6]
```diff
@@ -24,6 +24,18 @@ impl Direction2d {
...
+    pub unsafe fn new_unchecked(value: Vec2) -> Self {
```
Comment (james7132) [inline crates/bevy_math/src/primitives/dim2.rs:33]: This presents nothing that is actually undefined behavior in Rust. We should not be making APIs like these.

**Review by james7132 — CHANGES_REQUESTED** [review]

**Comment by doonv** [comment]
> Conventionally, `unsafe` in Rust is used strictly for memory safety issues.

Yes, but it's also used in cases where there is a promise that a certain type must have some limitation (being normalized in this case).

All the `NonZero` number types in the standard library of examples of this.

Having a direction vector that is not normalized may not cause UB in the traditional sense, but it could cause unforeseen bugs.

**Comment by tguichaoua** [comment]
> All the `NonZero` number types in the standard library of examples of this.

`NonZero` is used for niche optimization; if you create a zeored 'NonZero', then some code becomes unsound.
The question is: is the invariant of `Direction` (i.e., a vector is normalized) used as a guarantee in unsafe code?

**Comment by doonv** [comment]
> The question is: is the invariant of Direction (i.e., a vector is normalized) used as a guarantee in unsafe code?

No, but it almost certainly cause bugs in code that expect `Direction` to be normalized.

Marking it as `unsafe` makes the user think twice on if their vector is really normalized.

**Comment by tguichaoua** [comment]
From the [Book](https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html#unsafe-superpowers):
>  the intent is that as the programmer, you’ll ensure the code inside an unsafe block will access memory in a valid way.

`unsafe` in Rust is a matter of accessing valid memory. The worst that could happen with a non-normalized `Direction` is panic; which is a logic bug like dividing by `0`.

**Comment by doonv** [comment]
> The worst that could happen with a non-normalized Direction is panic

Panics are arent much of a problem here (they can easily be tracked down), what is a problem is a silent bug (A bug that doesn't provide any debug information when it occurs). For example, if you have a player movement function that takes in a `Direction2d` and that `Direction2d` isnt normalized, that cause the player to move slower/faster than expected, and it'll be very hard to track down. (The user will expect `Direction2d` to be normalized)

The `unsafe` keyword isn't just for memory issues and UB, it's for places that have a "contract" that the compiler can't check.

> The `unsafe` keyword has two uses:
> 
> - to declare the existence of contracts the compiler can't check (`unsafe fn` and `unsafe trait`),
> - and to declare that a programmer has checked that these contracts have been upheld (`unsafe {}` and `unsafe impl`, but also unsafe fn -- see below).

\- The `unsafe` keyword's documentation when hovered over (rust-analyzer) 

The "contract" here is that the inner vector is normalized.

As I said earlier, making this `unsafe` makes the user think twice about if their direction is really normalized. And it makes shooting yourself in the foot harder.

**Comment by nicopap** [comment]
This has been a longstanding discussion in the rust community. "What should I mark as unsafe?" For the standard library, the answer is "only methods with invariants, which, if not upheld, invoke undefined behaviors".

There is even a chapter on the rust reference about "behaviors not considered unsafe": <https://doc.rust-lang.org/stable/reference/behavior-not-considered-unsafe.html>. "Logic Errors" is in there.

This doesn't answer the question of whether we should mark the method as unsafe. The thing is, we already expose API that assumes the values passed-in represent a normal vector: [The `Quat` constructors](https://docs.rs/bevy/latest/bevy/math/struct.Quat.html#method.from_xyzw). I think we should be consistant with our API, and it makes sense to not mark the constructor as `unsafe`, the same way `glam` doesn't. We could add a runtime check though!

**Comment by doonv** [comment]
Well, I guess I'll make it safe.

File: crates/bevy_math/src/primitives/dim2.rs [hunk 7]
```diff
@@ -24,6 +24,18 @@ impl Direction2d {
         Self::new_and_length(value).map(|(dir, _)| dir)
     }
 
+    /// Create a [`Direction2d`] from a [`Vec2`] that is already normalized.
+    ///
+    /// ## Warning
```
Comment (tguichaoua) [inline crates/bevy_math/src/primitives/dim2.rs:29]:
```suggestion
    /// # Warning
```

**Review by tguichaoua — CHANGES_REQUESTED** [review]

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs [hunk 8]
```diff
@@ -166,7 +166,8 @@ impl Bounded3d for Capsule {
     fn aabb_3d(&self, translation: Vec3, rotation: Quat) -> Aabb3d {
         // Get the line segment between the hemispheres of the rotated capsule
         let segment = Segment3d {
-            direction: Direction3d::from_normalized(rotation * Vec3::Y),
+            // Multipling a normalized vector (Vec3::Y) with a rotation returns a normalized vector.
```
Comment (Jondolf) [inline crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs:169]:
```suggestion
            // Multiplying a normalized vector (Vec3::Y) with a rotation returns a normalized vector.
```

File: crates/bevy_math/src/primitives/dim3.rs [hunk 9]
```diff
@@ -146,8 +151,10 @@ impl Segment3d {
     pub fn from_points(point1: Vec3, point2: Vec3) -> (Self, Vec3) {
         let diff = point2 - point1;
         let length = diff.length();
+
         (
-            Self::new(Direction3d::from_normalized(diff / length), length),
+            // We are dividing by the length here, so the Vector is normalized.
```
Comment (Jondolf) [inline crates/bevy_math/src/primitives/dim3.rs:156]:
```suggestion
            // We are dividing by the length here, so the vector is normalized.
```

**Review by Jondolf — APPROVED** [review]
Some comment nits, but otherwise looks good

**Comment by rlidwka** [comment]
> The thing is, we already expose API that assumes the values passed-in represent a normal vector: [The Quat constructors](https://docs.rs/bevy/latest/bevy/math/struct.Quat.html#method.from_xyzw).

Quaternions do not need to be normalized. Only quaternions that are used for rotation do. I've used quaternions to represent angular velocity before, and those were intentionally not normalized.

In nalgebra, there's UnitQuaternion specifically for rotations, but there's nothing similar in bevy/glam.

**Pull request merged**